text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+- Framework versions might require an upgrade of the CLI. In which case, we provide instructions as to which CLI version should be used.
+
+This RFC details the necessary strategy and mechanism for ensuring that we keep those promises when releasing new versions.
+
+# Motivation
+
+There are 2 main motivations why we should put some substantial effort into this:
+
+## (1) Haven't we suffered enough?
+
+Recently, we experienced several incidents that are directly related to this domain:
+
+- https://github.com/aws/aws-cdk/issues/5986
+- https://github.com/aws/aws-cdk/issues/4897
+- https://github.com/aws/aws-cdk/issues/4294
+
+These issues were very awkward and caused both our customers and ourselves a lot of unwanted work.
+They also reflect poorly on the stability of the CDK, especially in a post GA era.
+
+*We want to implement this RFC to prevent (or at least greatly reduce) the occurrences of such issues.*
+
+## (2) Refactor CX Protocol versioning
+
+We've talked about the mechanism we currently use to perform protocol version changes quite a bit. The process and the expected output aren't very clear, and we seem to be always trying to understand it again and again.
+This might warrant a re-think/re-design, to make it much more ergonomic and developer friendly. 
+
+However, performing changes on such a crucial part of the system, shouldn't be done without a proper validation mechanism.
+
+*We want to implement this RFC so that we are able to iterate and change the versioning mechanism itself* 
+
+Actually, We have already acknowledged the need for this a while back:
+
+- https://github.com/aws/cdk-ops/issues/365
+- https://github.com/aws/cdk-ops/issues/233
+
+Its time we solve this once and for all.
+
+After implementing this RFC, we can expect the following outcomes:
+
+- An incompatible change in the CLI won't make it to the release.
+- An incompatible change in the framework, will always be accompanied with a version bump.
+- An incompatible change in the framework, accompanied with a protocol version bump, will always produce the correct error message.
+
+# Design Summary
+
+In order to ensure compatibility between the CLI and the Framework, we have to actually run the CLI against an old framework version, and vice versa.
+We want to find a generic mechanism for running these tests, without having to cherry-pick use-cases or versions. 
+
+The main design idea is to introduce new types of integration tests into our pre-release suite. Those tests will run various permutations of `(CLI X Framework)`, and validate the expected result in each permutation.
+
+To make it generic, the tests will run the entire current integration tests suite, and simply inject different versions each time.
+
+An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implement in older versions.
```

Comment:
I worry of what'll happen if there is an *intentionally* breaking change around some integration test that is *un-related* to the CX protocol? Running the old tests with the new code is bound to fail in this case, and the error message shouldn't be expected to be that of a CX protocol change...?

Reply:
```suggestion
An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implemented in older versions.
```

Reply:
Thats an excellent point. Basically, my reasoning around this is that if a scenario like that occurs, it might mean the integration test isn't actually testing the right thing. 

It feels like integration tests should only validate functionality we shouldn't be breaking (cdk command line options, docker build command, and other user facing behavior). 

There are definitely cases where we "break" some behavior, that isn't necessarily impacting users that much, which is why its ok to do. 

For example, the [simplify docker publishing](https://github.com/aws/aws-cdk/pull/5733) PR removed the `ECRAdoptedRepository` construct, which is technically a breaking change. The failing tests due to that, are unit tests, not integration tests (need to double check). My argument is that if we did have an integration test that validates the existence of an `ECRAdoptedRepository`, that validation probably didn't belong their in the first place. In essence, it means the integration test was validating the implementation rather than the expected result, which is that images are properly treated as assets (uploaded to somewhere you can download it from). 

This assumption might be wrong. Can you think of an integration test that will fail rightfully fail if we introduce such "breaking" changes? or even a theoretic scenario?

Reply:
The questions isn't so much about whether those are "valid" breakages or not... It's more about how do you move past those in the pipeline? If we automatically run the **previous release's** integration tests, there is no way for us to actually amend the "invalid" tests there...

Having to resort to manually releasing software doesn't sound like a safe thing to do...

Reply:
@RomainMuller  see https://github.com/aws/aws-cdk-rfcs/blob/epolon/compatibility-strategy/text/00110-cli-framework-compatibility-strategy.md#exclusions